File: src/main/java/gargoyle/netsquares/prop/simple/StringProperty.java
package gargoyle.netsquares.prop.simple;

import gargoyle.netsquares.prop.RWProperty;

import java.io.Serializable;
import java.util.Objects;

public final class StringProperty extends RWProperty<String> implements Comparable<StringProperty>, Serializable {
    public StringProperty() {
        this("");
    }

    public StringProperty(String value) {
        super(value);
    }

    @Override
    public int compareTo(StringProperty o) {
        String thisValue = get();
        String thatValue = o.get();
        return (thisValue == null ? "" : thisValue).compareTo(thatValue == null ? "" : thatValue);
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (!(o instanceof RWProperty)) return false;
        if (!super.equals(o)) return false;
        RWProperty<?> that = (RWProperty<?>) o;
        return Objects.equals(get(), that.get());
    }
}

File: src/main/kotlin/gargoyle/netsquares/prop/simple/StringProperty.kt
package gargoyle.netsquares.prop.simple

import gargoyle.netsquares.prop.RWProperty

class StringProperty constructor(value: String = "") : RWProperty<String>(value),
    Comparable<StringProperty> {
    override fun compareTo(other: StringProperty): Int {
        val thisValue = get()
        val thatValue = other.get()
        return thisValue.compareTo(thatValue)
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is RWProperty<*>) return false
        if (!super.equals(other)) return false
        return get() == other.get()
    }

    override fun hashCode(): Int = get().hashCode()
}
